flytekit/common/workflow.py

```diff
@@ -171,33 +213,39 @@ def fetch(cls, project, domain, name, version=None):
         return sdk_workflow
 
     @classmethod
-    def promote_from_model(cls, base_model):
+    def promote_from_model(cls, base_model, sub_workflows=None, tasks=None):
         """
         :param flytekit.models.core.workflow.WorkflowTemplate base_model:
+        :param list[flytekit.models.core.workflow.WorkflowTemplate] sub_workflows: Provide a list of WorkflowTemplate
+            models (should be returned from Admin as part of the admin CompiledWorkflowClosure. Relevant sub-workflows
+            should always be provided.
+        :param list[flytekit.models.task.TaskTemplate] tasks: Same as above but for tasks. If tasks are not provided
+            relevant TaskTemplates will be fetched from Admin
         :rtype: SdkWorkflow
         """
-        node_map = {
-            n.id: _nodes.SdkNode(
-                n.id,
-                [],
-                n.inputs,
-                n.metadata,
-                sdk_task=None,
-                sdk_workflow=None,
-                sdk_branch=None  # TODO: Hydrate these objects by reference from the engine.
-            )
-            for n in base_model.nodes
-        }
-
-        for v in _six.itervalues(node_map):
-            v.upstream_nodes[:] = [node_map[k] for k in v.upstream_node_ids]
-
+        node_map = {}
+        for n in base_model.nodes:
+            if n.id == _constants.START_NODE_ID or n.id == _constants.END_NODE_ID:
+                # The workflow compilation process done by Admin/Propeller will add a fake start-node to the graph
+                # so we need to strip them back out here.
+                continue
+            else:
+                node_map[n.id] = _nodes.SdkNode.promote_from_model(n, sub_workflows, tasks)
+
+        # Set upstream nodes for each node
+        for n in base_model.nodes:
+            if n.id == _constants.START_NODE_ID or n.id == _constants.END_NODE_ID:
```

Comment:
nit: put everything in a conditional like `if n.id not in {START_N)DE_ID, END_NODE_ID}`...and maybe make that a function just in case we add more special case nodes in the future.

Reply:
or pre-prune out the system nodes.

Reply:
https://github.com/lyft/flytekit/pull/75/commits/f499dd1c5c6dfaefd539d964ed86b47a7ea7794f please check @matthewphsmith